### PR TITLE
chore(external docs): Update docs to reflect acknowledgements support

### LINKED ISSUE #7412: Add documentation for end-to-end acknowledgement to the reference
The current support for end-to-end acknowledgements has been introduced without any external reference documentation. This needs to be completed to note which sinks and sources have support, and how it changes Vector's behavior.

### PR DESCRIPTION
**Pull request by bruceg**
Closes #7412 

Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

### CODE REVIEW

File: docs/manual/about/under-the-hood/guarantees.md
Comment (blt) [inline docs/manual/about/under-the-hood/guarantees.md:27]:
```suggestion
For a source, this means that the component is able at the protocol
level to both indicate to the sender of the events a status code when and if
```

**Review by blt — APPROVED** [review]
One very small wording shuffle nit, not blocking.

File: docs/reference.cue
Comment (binarylogic) [inline docs/reference.cue:24]: After thinking about this more, I'd prefer that we unify`#AcknowledgementsStatus` and `#DeliveryStatus`. I’m of the mindset that a component is `at_least_once` possible if it has both:

1. An underlying protocol that acknowledges receipt
2. Implements Vector-level acknowledgements

If either are violated then the component is `best_effort`, otherwise the guarantee is satisfied from a user's perspective.

cc @lucperkins

File: docs/manual/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md:27]: I think this changed version isn't quite accurate. The status code indication happens at the protocol level, while the source waiting happens within vector. Re-reading it, the original is poorly worded as well. I'll try again.

File: docs/reference.cue
Comment (bruceg) [reply docs/reference.cue:24]: It seems somewhat reductive to boil it all down to just two states, but I think I see what you're getting at. This will of course take the `at_least_once` state away from most of the components (for now), but that may well have been misleading anyways.

Comment (bruceg) [reply docs/reference.cue:24]: How should this work for sources that scrape metrics? They can't really be `at_least_once`, but at least some of the sources like this are marked that way.

File: docs/manual/about/under-the-hood/guarantees.md
Comment (lucperkins) [inline docs/manual/about/under-the-hood/guarantees.md:23]: `at_least_once` (to be consistent with how it will appear in configs)

**Review by lucperkins — APPROVED** [review]
One small nit in the docs and otherwise good to go

File: docs/manual/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md]: Fixed (I think)

Comment (lucperkins) [reply docs/manual/about/under-the-hood/guarantees.md]: Oh sorry, I meant just in the code word, not in the headings

**Comment by bruceg** [comment]
@lucperkins are you good with the latest `at-least-once` change?

File: docs/manual/about/under-the-hood/guarantees.md
Comment (StephenWakely) [inline docs/manual/about/under-the-hood/guarantees.md:25]: This may be pedantic, but presumably we can't guarantee that it will be delivered, but we can guarantee that if it isn't delivered the source will return an error? Eg, if I try to push to the Datadog sink but I don't have an account, it will never be delivered, despite our guarantee. What happens in that situation?

Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md:25]: There are two senses to the term "delivered", both an attempt to send the event and the confirmation of receipt. This is trying to explain that the sink will attempt delivery until a confirmation or error is received, and that sources will wait for that confirmation before acknowledging. I agree the wording is less than ideal, so I welcome better suggestions.

**Review by blt — APPROVED** [review]

File: docs/manual/about/under-the-hood/guarantees.md
Comment (binarylogic) [inline docs/manual/about/under-the-hood/guarantees.md:26]:
```suggestion
for _all_ connected sinks to mark the event as delivered before
```

File: docs/manual/about/under-the-hood/guarantees.md
Comment (binarylogic) [inline docs/manual/about/under-the-hood/guarantees.md:38]: Is this still true? If ack'ing is enable, and Vector dies while an event is in-flight, it won't be acknowledged, right?

File: docs/manual/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md:26]: Yeah, that's worth emphasizing.

File: docs/manual/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md:38]: For those sinks that support it, yes, so this section needs to be adjusted to reflect that.

File: docs/manual/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline docs/manual/about/under-the-hood/guarantees.md:30]: This isn't the case when buffers are used, right? It will acknowledge once it has been written to the buffer?

File: docs/reference/components/sinks/aws_cloudwatch_logs.cue
Comment (jszwedko) [inline docs/reference/components/sinks/aws_cloudwatch_logs.cue:8]: 🤔 There is some nuance here that it is only `at_least_once` when disk buffers are used, but I think this and many of the below components are still `at_least_once`, no? I think only protocols that don't support acknowledgement, like UDP, would be `best_effort`. This is mentioned in the above guarantees.md:

> In order to achieve at least once delivery between restarts your sink must be configured to use `disk` based buffers:

End-to-end acknowledgements change this, but at a topology level rather than a component level, I think.

File: docs/manual/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/manual/about/under-the-hood/guarantees.md:30]: Correct.

File: docs/reference/components/sinks/aws_cloudwatch_logs.cue
Comment (bruceg) [reply docs/reference/components/sinks/aws_cloudwatch_logs.cue:8]: Before this change, the `at_least_once` indicated simply that we would retry until the events were accepted or explicitly rejected. Now, it has been overloaded to also indicate that it participates in end-to-end acknowledgements. I am open to a better solution to this, as I don't particularly like this overloading, but see also this: https://github.com/timberio/vector/pull/7436#discussion_r634764018

Comment (jszwedko) [reply docs/reference/components/sinks/aws_cloudwatch_logs.cue:8]: Agreed, I think we'll need to come up with some better terms and definitions as part of this. We can delay thinking about it too much until we pick up the work to officially release end-to-end acknowledgements though.

**Comment by bruceg** [comment]
Ok, this is rewritten and ready for re-review. Note that many of the sinks' delivery class will be upgraded once #8426 lands, so I'll update the PRs as needed when one of these is merged.

File: docs/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline docs/content/en/docs/about/under-the-hood/guarantees.md:28]: Should we add that the guarantee only applies if end-to-end acknowledgements are enabled?

File: docs/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline docs/content/en/docs/about/under-the-hood/guarantees.md:21]: I think we are still missing the nuance here that, if a disk buffer is configured on the sink, that it won't wait for the event to be propagated all the way to where-ever the sink is sending, but just until it is persisted in the disk buffer.

I think this nuance is important for users to be aware of. They may naively think that the source won't ack until the event is literally pushed downstream of Vector.

**Review by jszwedko — COMMENTED** [review]
Thanks for the updates here. I think it reads more clearly now. I left a couple of other notes.

I think we should hold off on merging this until we officially release the end-to-end acknowledgement feature though. Until we've done more testing of it to understand how it affects throughput and suss out any bugs, I hesitate to push people to use it.

File: docs/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/content/en/docs/about/under-the-hood/guarantees.md:28]: Ah yes, agreed.

File: docs/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply docs/content/en/docs/about/under-the-hood/guarantees.md]: Does 3790d24e3 work for you?

Comment (jszwedko) [reply docs/content/en/docs/about/under-the-hood/guarantees.md]: That works, thanks!

**Comment by jszwedko** [comment]
I believe this work will be picked up this week as Bruce is jumping back into the end-to-end ack work.

**Comment by bruceg** [comment]
I think this was just waiting for us to be ready to announce the feature. Could I get a re-review @lucperkins @binarylogic @jszwedko

**Comment by bruceg** [comment]
This is ready for a re-review so we can merge it and keep it up to date.

**Review by blt — APPROVED** [review]

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (tobz) [inline website/content/en/docs/about/under-the-hood/guarantees.md:22]: I would change this to `durable buffer`.  In the future, external buffer implementations that provide durability as a first-class citizen will also acknowledge events in the same way as disk buffers.

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (tobz) [inline website/content/en/docs/about/under-the-hood/guarantees.md:33]: Comma between `restarts` and `your`.

**Review by tobz — CHANGES_REQUESTED** [review]
Two small technical nits.

**Review by tobz — APPROVED** [review]

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline website/content/en/docs/about/under-the-hood/guarantees.md:26]: This was noted offline, but I think this conflates the transport delivery guarantees with acknowledgements. We mentioned having a separate field for sinks indicating acknowledgement support.

**Review by jszwedko — CHANGES_REQUESTED** [review]
Just one comment, otherwise looks good!

**Review by binarylogic — APPROVED** [review]
Nice work

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline website/content/en/docs/about/under-the-hood/guarantees.md:14]: I think we should add a note to this section about how you can determine if the source and sink you are using support acknowledgements and what happens if the sink does not support them.

**Review by jszwedko — COMMENTED** [review]
Just one last note, otherwise I think this is good.

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply website/content/en/docs/about/under-the-hood/guarantees.md:14]: Done in 3aa48fabb

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (tshepang) [inline website/content/en/docs/about/under-the-hood/guarantees.md:18]:
```suggestion
an event through the lifetime of that event as it travels from the
```

**Comment by bruceg** [comment]
Blocked by #11615

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply website/content/en/docs/about/under-the-hood/guarantees.md]: Fixed, thanks.

**Comment by bruceg** [comment]
This should be ready for final review, all outstanding issues are resolved AFAICT.

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (tobz) [inline website/cue/reference/components/sinks/splunk_hec.cue:8]: Why is this going away?

File: website/cue/reference/components/sources/kubernetes_logs.cue
Comment (tobz) [inline website/cue/reference/components/sources/kubernetes_logs.cue:23]: Hmm, really? Even though it uses the `file` source machinery under-the-hood?

File: website/cue/reference/components/sources/syslog.cue
Comment (tobz) [inline website/cue/reference/components/sources/syslog.cue:11]: Seems like this should be hard-coded to `false`? I can't conceive how we'd accomplish a protocol-oblivious means of doing acknowledgements for `socket`-based sources.

**Review by tobz — COMMENTED** [review]
A few nits/questions.

File: website/cue/reference/components/sinks/splunk_hec.cue
Comment (bruceg) [reply website/cue/reference/components/sinks/splunk_hec.cue:8]: It conflicts with/shadows the global `acknowledgements` options, and that breaks validation.

File: website/cue/reference/components/sources/syslog.cue
Comment (bruceg) [reply website/cue/reference/components/sources/syslog.cue:11]: Sure. I basically just copied the pattern of the other features here.

File: website/cue/reference/components/sources/kubernetes_logs.cue
Comment (tobz) [reply website/cue/reference/components/sources/kubernetes_logs.cue:23]: TIL.

File: website/cue/reference/components/sources/syslog.cue
Comment (tobz) [reply website/cue/reference/components/sources/syslog.cue:11]: Yeah, understandable.  It's probably fine as-is and we'll hopefully get to the configuration schema to generate this stuff before it would ever matter, again, given how unlikely it is that we ever somehow were able to add generic ack support to the `socket` source. 😅

**Review by tobz — APPROVED** [review]

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline website/content/en/docs/about/under-the-hood/guarantees.md:21]: I'm not actually seeing the badges. I think we may need to update `website/layouts/partials/docs/component-under-hero.html`?

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline website/content/en/docs/about/under-the-hood/guarantees.md:24]: Is this true? I thought we moved the `acknowledgements` option to the sinks?

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (jszwedko) [inline website/content/en/docs/about/under-the-hood/guarantees.md:40]: One question that came up recently, that we may want to get ahead of by documenting here, is what happens if events are dropped somewhere in the topology? My understanding is that they will be "acknowledged" by the source.

**Review by jszwedko — COMMENTED** [review]
Just a couple of notes, but otherwise 👍

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply website/content/en/docs/about/under-the-hood/guarantees.md:24]: It is still true, but incomplete and deprecated. I will reword this.

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply website/content/en/docs/about/under-the-hood/guarantees.md:40]: Correct. I will add a note about that.

File: website/content/en/docs/about/under-the-hood/guarantees.md
Comment (bruceg) [reply website/content/en/docs/about/under-the-hood/guarantees.md:21]: Ah, yes, I had been looking for that, but understood from another comment that the badges were found elsewhere. Fixed in e4878bf5f (I hope)

Comment (jszwedko) [reply website/content/en/docs/about/under-the-hood/guarantees.md:21]: 👍 I see the badges now.

**Review by jszwedko — APPROVED** [review]
A couple of last small edits, otherwise looks good!

One other suggestion I have is to document the behavior of fanout, if one source is connected to multiple sinks, where only one of them has acknowledgements enabled.

**Pull request merged**
